### PR TITLE
fix(spooler): make Iot limit size and publish retry configurable and …

### PR DESCRIPTION
**Pull request by awszztt**
…validate the publish reqeust

**Issue #, if available:**

**Description of changes:**

**Why is this change necessary:**

**How was this change tested:**

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [ ] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/mqttclient/MqttClient.java [hunk 1]
```diff
@@ -293,6 +313,25 @@ protected MqttClient(DeviceConfiguration deviceConfiguration, Spool spool, boole
...
+    }
+
+    private void validateAndSetMqttPublishConfiguration() {
+        maxPublishMessageSize = Coerce.toInt(mqttTopics.findOrDefault(DEFAULT_MQTT_MAX_OF_MESSAGE_SIZE_IN_BYTES,
+                MQTT_MAX_OF_MESSAGE_SIZE_IN_BYTES_KEY));
+        if (maxPublishMessageSize > MQTT_MAX_LIMIT_OF_MESSAGE_SIZE_IN_BYTES) {
+            logger.atWarn()
+                    .kv(MQTT_MAX_OF_MESSAGE_SIZE_IN_BYTES_KEY, maxPublishMessageSize)
+                    .kv("Max acceptable configuration", MQTT_MAX_LIMIT_OF_MESSAGE_SIZE_IN_BYTES)
+                    .log(String.format("The configuration of %s exceeds the max limit and will change to "
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqttclient/MqttClient.java:326]: don't use string.format in a log. Use the built in interpolation using `{}`

File: src/main/java/com/aws/greengrass/mqttclient/MqttClient.java [hunk 2]
```diff
@@ -455,12 +502,53 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
...
+
+        // Topic should not contain wildcard characters
+        String topic = request.getTopic();
+        if (topic.contains("#") || topic.contains("+")) {
+            throw new SpoolerStoreException("The topic of publish request should not contain wildcard "
+                    + "characters of '#' or '+'");
+        }
+
+        String reservedTopicTemplate = "^\\$aws/rules/\\S+/\\S+";
+        String prefixOfReservedTopic = "^\\$aws/rules/\\S+?/";
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqttclient/MqttClient.java:521]: pull these out as constants.

**Review by MikeDombo — APPROVED** [review]
LGTM, just make sure the build passes

File: src/integrationtests/java/com/aws/greengrass/integrationtests/mqttclient/MqttClientPublishTest.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+public class MqttClientPublishTest extends BaseITCase {
+
+    private static Kernel kernel;
+    private static EventStreamRPCConnection clientConnection;
+    private static SocketOptions socketOptions;
+    private static final byte[] TEST_GOOD_PAYLOAD = "goodPayload".getBytes(StandardCharsets.UTF_8);
+    private static final int TIMEOUT_FOR_MQTTPROXY_SECONDS = 20;
+    private static final String TEST_GOOD_PUBLISH_TOPIC = "A/B/C";
+    private static ExecutorService executorService = TestUtils.synchronousExecutorService();
+    private static Configuration config = new Configuration(new Context());
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/mqttclient/MqttClientPublishTest.java:77]: the `kernel` that you have already has a `Configuration` do you really need to make another one?

File: src/main/java/com/aws/greengrass/mqttclient/MqttClient.java [hunk 4]
```diff
@@ -455,12 +511,51 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
...
+                    + "characters of '#' or '+'");
+        }
+
+        if (Pattern.matches(reservedTopicTemplate, topic.toLowerCase())) {
+            // remove the prefix of "$aws/rules/rule-name/"
+            topic = topic.toLowerCase().split(prefixOfReservedTopic, 2)[1];
+        }
+
+        // Topic should not have no more than maximum number of forward slashes (/)
+        if (topic.chars().filter(num -> num == '/').count() > MAX_NUMBER_OF_FORWARD_SLASHES) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqttclient/MqttClient.java:535]: this limit should also be applied to subscriptions, right?

**Review by MikeDombo — APPROVED** [review]

**Review by MikeDombo — DISMISSED** [review]

File: src/main/java/com/aws/greengrass/mqttclient/PublishRequest.java [hunk 5]
```diff
@@ -13,7 +13,7 @@
 
 @Builder
 @Value
-public class PublishRequest {
+public class PublishRequest extends MqttRequest {
     @NonNull String topic;
```
Comment (popanmol) [inline src/main/java/com/aws/greengrass/mqttclient/PublishRequest.java:17]: MqttRequest already has the field topic, so shouldn't need to declare it here again. Same for subscribe and unsubscribe

File: src/main/java/com/aws/greengrass/mqttclient/MqttRequest.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.experimental.FieldDefaults;
+import lombok.experimental.SuperBuilder;
+
+@SuperBuilder
+@Getter
+@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+public class MqttRequest {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqttclient/MqttRequest.java:17]: I think this type is overkill. Just pass in the topic directly to the validation method.

**Review by MikeDombo — DISMISSED** [review]

**Review by popanmol — APPROVED** [review]

**Pull request merged**